Add enhancement proposal for generic constraints feature (#91)
operator-framework/enhancements: enhancements/generic-constraints.md

## Motivation

At the moment, OLM only supports 2 built-in types of constraint:
- GVK constraint (`olm.gvk.required`)* which depends on a specific GroupVersionKind information of a CRD/API.
- Package constraint (`olm.package.required`)* which depends a specific package name and version (or a range of version)

Both types are specifically handled by the resolver in OLM and cannot be changed or expanded to support further use cases. In order to support more types of constraints, it is generically expected for OLM to add more specific built-in types to allow users to express those new constraints in the bundles. Over the time, it may become cumbersome and impractical to continue to add more specific types to the resolver. The need for OLM to support a generic constraint model becomes more apparent. With a generic constraint model, OLM no longer needs to carry any specific evaluation methodology besides the understanding of how to parse the constraint syntax. The users should be able to express declaratively the rules for a constraint that can be evaluated againts the dataset of arbitrary properties.

> * Note: The current GVK and package constraints are specified in `dependencies.yaml` using `olm.gvk` and `olm.package` type and they are coverted to `olm.gvk.required` or `olm.package.required` property (which is required constraint) to differentiate from the `olm.package` and `olm.gvk` property in the bundle.

Reviewer: ```suggestion
* Note: The current GVK and package constraints are specified in `dependencies.yaml` using `olm.gvk` and `olm.package` type and they are converted to `olm.gvk.required` or `olm.package.required` property (which is required constraint) to differentiate from the `olm.package` and `olm.gvk` property in the bundle.
```

The CEL library also supports [extension functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions) which allows OLM to implement additional functions that may be necessary for certain operations. These functions then can be used within the CEL expression. For example, [semver](https://semver.org/) comparison is not built-in in CEL and it is often used in OLM as bundles are versioned in semver format. The semver functions can be added in the initial release and allow users to express semver evaluations in the constraint. However, extension functions should be treated and designed with care as they can be difficult to change in the future and potentially create version skew or breaking changes.

> All CEL expressions are [parsed and typechecked](https://github.com/google/cel-go#parse-and-check) before evaluation. The insulting programs can be cached for later evaluation to reduce unnecesary computation. Also, the CEL evaluation is [thread-safe and side-effect free](https://github.com/google/cel-go#evaluate).

Reviewer: ```suggestion
All CEL expressions are [parsed and typechecked](https://github.com/google/cel-go#parse-and-check) before evaluation. The resulting programs can be cached for later evaluation to reduce unnecessary computation. Also, the CEL evaluation is [thread-safe and side-effect free](https://github.com/google/cel-go#evaluate).
```

#### Compound Constraint

As an operator author, I would like to specify a constraint that has multiple rules and requirements that must be resolved into a single dependent operator. For example, I would like to have a constraint for a dependent operator that belongs to a specific package and provides a specific API/GVK.

### Design Details

#### Generic Constraint Syntax

The constraint syntax is designed to have similarity with the existing constraints that OLM supports which are GVK and package. It is intentional to keep the fundamental syntax with `type` and `value` the same so that it doesn't introduce any dramatic changes to the overall user experiences.

Each constraint is specified using the following syntax:

```json=
{
   "type":"olm.constraint",
   "value":{
      "evaluator":{
         "id":"cel"
      },
      "rule": "properties.exists(p, p.type == \"certified\")",
      "message": "require to have certified property",
      "action":{
         "id":"require"
      }
   }
}
```

The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 4 fields: `evaluator`, `rule`, `message` and `action`.

> The `evaluator` is a struct with `id` field to represent the language library that will be used to evaluate the expression. At the moment, only CEL expression is supported using `cel` as the identifier.

Reviewer 2: What's the purpose of `evaluator` being an object with an `id` field rather than having `evaluator` being the string ID directly?

Do we envision needing to add other fields to the evaluator object? If so, what are some examples of those other fields that we know we'll need in the future?

Reviewer: I'd go further: why not omit the `evaluator` entirely and identify via the type?

```json
"type":"olm.constraint.cel",
```

Same for action:
```json
{
   "type":"olm.constraint.cel.required",
   "value": "properties.exists(p, p.type == \"certified\")",
}
```

Edit: 

Just noticed this form in the section at the end. Any reason it's not the preferred form?

Reviewer 2: My concern is that different root `type`s will cause version skew concerns when we add a new constraint type. 

If we:
- add support for `olm.constraint.cel.required` in 0.20
- add support for `olm.constraint.foobar.required` in 0.21

And then a bundle contains `olm.constraint.foobar.required` and is attempted to be installed in <=0.20, what does OLM do with that property?

In 0.20, support for that as a known constraint is not present, so OLM would treat it like an opaque property and ignore it entirely. There's no opportunity to fail resolution (if its actually required) or ignore it on purpose (because its optional) because OLM can't tell user intent since we're co-mingling constraints and properties.

> The `rule` field is the string that presents a CEL expression that will be evaluated during resolution against . Only CEL expression is supported in the initial release.

Reviewer 2: Does this field apply only when `evaluator.id` is `"cel"`? If not, it we should describe the general purpose of this field and have a few examples (one of which should be a cel rule).

> 

Reviewer 2: Separate thread about `rule`... 
- Can you go into some detail about what variables will be available in the context of an expression? Is it just properties?
- Are we planning to enable [macros](https://github.com/google/cel-go#macros)? If so, all or a subset?
- Are we planning to implement any [custom functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions)? If so, what are they?

The `message` field is an optional field that is accommodating the rule field to surface information regarding what this CEL rule is about. The message will be surfaced in resolution output when the rule evaluates to false.

> Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. This action can be used to indicate optional constraint in the future. For the initial release, `require` action is supported.

Reviewer 2: Same question as I have for `evaluator`. Why do we need the sub `id` field? What are some examples of other fields that we know we'll need under the action object?

Will all actions be applicable to all future evaluators and all rules?

> 

Reviewer 2: General questions:
1. Which of these fields are required vs optional?
2. Are we expecting that _all_ future constraint types can be expressed by `olm.constraint`?
3. If I later want to add another native constraint (e.g. propertyExists), how would that look?

If the answers to 2) and 3) are "no, we'll make a separate root `type` (e.g. `olm.constraint.propertyExists`)", then what is the stance on version skew when a bundle contains a property (that's actually intended to be a constraint) that is unknown to OLM?

> #### Compound Constraint

Reviewer 2: Given that #97 focuses on general compound constraints, it seems like our stance should be to favor that as a best practice. While the approach described here is just as valid from a technical perspective, it seems like the approach in #97 integrates better into how the resolver works and will give users more insight into which aspect(s) of the compound constraint caused resolution failures.

Besides CEL, there are other languages that can be used such as:

- [Rego](https://github.com/open-policy-agent/opa/tree/main/rego) (Open Policy Agent)
- [Expr](https://github.com/antonmedv/expr)
- [Starlark](https://github.com/bazelbuild/starlark)
- [Cue](https://github.com/cue-lang/cue)

All of these languages can be supported in constraint type. The `evaluator` field is designed to support multiple evaluators/languages if needed. However, intrducing a new language to the constraint type should be evaluated carefully to ensure there is a real need for it. Providing the support for multiple languages can be overwhelming and potentially creates a fragmented user experiences and unnecessary maintainance effort in a long run.

### Simplified constraint type

The current proposed constraint syntax has most of information nested under `value` field. The complexity of a struct with nested fields can create a bad user experience. It is possible to introduce a constraint type that is specified for CEL

```json=
{
  "type": "olm.constraint.cel",
  "value": {
    "rule": 'properties.exists(p, p.type == "certified") && properties.exists(p, p.type == "stable")',

>     "message": 'require to have "certified" and "stable" properties',

Reviewer: This just seems like restating the constraint. Is this actually a field that should be available on in the property metaschema?

```
{
  "type": "proptype",
  "value": <>, 
  "failureMessage": "etc"
}
```

Reviewer 3: I'd buy `"description"` making sense for all properties.

Reviewer 2: Based on the current state of this EP, I think I'd vote to start with `failureMessage` in the root of the `olm.constraint` value struct.

A general `description` and a constraint `failureMessage` seem different enough semantically to me that I don't think we should overload one onto the other.

#### Compound Constraint

As an operator author, I would like to specify a constraint that has multiple rules and requirements that must be resolved into a single dependent operator. For example, I would like to have a constraint for a dependent operator that belongs to a specific package and provides a specific API/GVK.

### Design Details

#### Generic Constraint Syntax

The constraint syntax is designed to have similarity with the existing constraints that OLM supports which are GVK and package. It is intentional to keep the fundamental syntax with `type` and `value` the same so that it doesn't introduce any dramatic changes to the overall user experiences.

Each constraint is specified using the following syntax:

```json=
{

>    "type":"olm.constraint",

Reviewer 3: why are all types prefixed by `olm`? shouldn't it be `deppy` or maybe nothing?

Reviewer 2: I think we should define at least one reserved prefix and document that custom properties should also use different unique prefixes (maybe owned domain?) to ensure uniqueness across the global set of properties that might exist. Uniqueness is especially important now that we're introducing a constraint syntax that supports arbitrary expressions against arbitrary properties.

The `evaluator` is a struct with `id` field to represent the language library that will be used to evaluate the expression. At the moment, only CEL expression is supported using `cel` as the identifier. Given the possibility of supporting other expression languages in the future, the `evaluator` is constructed as a struct so that additional fields can be added without breaking the current syntax. For example, if we support a hypothetical expression language named `lucky` that has ability to dynamically load a package named `bobby` at runtime, a `package` field can be added to `evaluator` struct to support that ability:

```json=
"evaluator":{
    "id":"lucky",
    "package":"bobby"
},
```

The `rule` field is the string that presents a CEL expression that will be evaluated during resolution against . Only CEL expression is supported in the initial release.

The `message` field is an optional field that is accommodating the rule field to surface information regarding what this CEL rule is about. The message will be surfaced in resolution output when the rule evaluates to false.

> Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. This action can be used to indicate optional constraint in the future adding a new field `optional` such as:

Reviewer 4: I don't really understand an "optional constraint". I understand an optional dependency, meaning that an API / package may need to be installed if possible but not a "constraint". Would you mind clarifying what is behind it, giving an example would be great?

Reviewer 2: Outdated now since the action field is no longer in scope in this EP. However, for posterity...

An explicit package dependency is just a type of constraint.
- A required dependency looks like:  "A valid SAT solution MUST include a package that meets the dependency constraint"
- An optional dependency looks like: "A valid SAT solution CAN include a package that meets the dependency constraint and should be favored over a solution that does not"

Generalizing to arbitrary constraints, I think it would translate to:
- A required constraint looks like: "A valid SAT solution MUST include a package that meets the constraint"
- An optional constraint looks like: "A valid SAT solution CAN include a package that meets the constraint and should be favored over a solution that does not"

### Design Details

#### Generic Constraint Syntax

The constraint syntax is designed to have similarity with the existing constraints that OLM supports which are GVK and package. It is intentional to keep the fundamental syntax with `type` and `value` the same so that it doesn't introduce any dramatic changes to the overall user experiences. The proposed constraint syntax is designed to support other expression languages in the future. Additionally, other goals of this EP to support potential custom constraints which may not use expression language and have additional fields without having to introduce a new constraint type. As a result, each constraint type is designed to be an individual struct with custom fields that are specific to that type. New types can be introduced and added to `value` struct to support new constraint type without needing a new `type` identifier.

Each constraint is specified using the following syntax:

```yaml=
type: olm.constraint
value:
    message: 'require to have "certified" and "stable" properties'

>     CEL:

Reviewer 2: Nit, I'd make this all lowercase:
```suggestion
    cel:
```

Author: +1

```json=
{
  "type": "olm.constraint",
  "value": {
      "message": 'require to have "certified" and "stable" properties',
      "CEL": {
          "rule": 'properties.exists(p, p.type == "certified") && properties.exists(p, p.type == "stable")'
      }
  }
}
```

The expression in `rule` has two requirements that are linked in an AND (&&) logical operator. In order for that expression to be true, both requirements must be satisfied. Therefore, the resolved operator will be a single bundle that has both certified and stable properties. This concept represents the compound constraint.

### Graduation Criteria

This generic constraint is currently scheduled to be available in the next release with a set of supported functionalities. More functionalities will be added to the feature if needed in the subsequent releases.

Alternatively, it is possible to wrap this feature under feature flag which means it is disabled by default. As a result, the feature can be a dev-preview/tech-preview feature and begins to gather users' feedbacks.

> As the feature is being used and evaluated in production, changes can be made to address feedbacks and improve overall usefulness and user experience. Then, the feature flag can be removed and the feature is enabled by default moving forward. At the time, the feature is GA.

Reviewer 2: Did we make a final decision on tech preview vs GA?

Author: As far as I know, it is gonna be GA right away especially if we need it for the cluster constraint.

If CEL becomes too unstable in the future due to breaking changes, it is possible to gradually deprecate the support of the language and introduce another language that is more stable. The constraint syntax allows the support of other languages in the future via `evaluator` field.

## Alternatives

### Other languages choices

While CEL is the selected language to be supported in the new OLM constraint type. It is not the intention to limit OLM to only support CEL expression. It possible to support multiple expression languages in the future releases if needed.

Besides CEL, there are other languages that can be used such as:

- [Rego](https://github.com/open-policy-agent/opa/tree/main/rego) (Open Policy Agent)
- [Expr](https://github.com/antonmedv/expr)
- [Starlark](https://github.com/bazelbuild/starlark)
- [Cue](https://github.com/cue-lang/cue)

> All of these languages can be supported in constraint type. The `evaluator` field is designed to support multiple evaluators/languages if needed. However, introducing a new language to the constraint type should be evaluated carefully to ensure there is a real need for it. Providing the support for multiple languages can be overwhelming and potentially creates a fragmented user experiences and unnecessary maintenance effort in a long run.

Reviewer 2: Update this paragraph now that each constraint implementation gets its own config struct?

Author: +1
